src/requests/create-signature.ts

```diff
@@ -0,0 +1,107 @@
+import * as crypto from 'crypto'
+import type {
+  CanonicalRequest,
+  NormalizedCanonicalRequest,
+  NormalizedHeaders,
+  Secret,
+  Timestamp,
+} from './typings'
+import { CanonicalRequestValidator, SecretValidator, TimestampValidator } from './typings'
+import { getNormalizedEncodedURI, getNormalizedHeaders } from './utils'
+import { ContentfulSigningHeader } from './constants'
+
+const hash = (normalizedCanonicalRequest: NormalizedCanonicalRequest, secret: string) => {
+  const stringifiedHeaders = normalizedCanonicalRequest
+    .headers!.map(([key, value]) => `${key}:${value}`)
+    .join(';')
+
+  const stringifiedRequest = [
+    normalizedCanonicalRequest.method,
+    normalizedCanonicalRequest.path,
+    stringifiedHeaders,
+    normalizedCanonicalRequest.body,
+  ].join('\n')
+
+  const hmac = crypto.createHmac('sha256', secret)
+
+  hmac.update(stringifiedRequest)
+
+  return hmac.digest('hex')
+}
+
+const enrichNormalizedHeadersWithMetadata = (headers: NormalizedHeaders, timestamp: number) => {
+  const result = [...headers]
+  const joinedSignedHeaders = headers.map(([key]) => key).join(',')
+
+  result.push([ContentfulSigningHeader.Timestamp, timestamp.toString()])
+
+  if (joinedSignedHeaders) {
+    result.push([ContentfulSigningHeader.SignedHeaders, joinedSignedHeaders])
```

Comment:
shouldn't `x-contentful-timestamp,x-contentful-signed-headers` be always on the list? we should be signing them too

Reply:
Timestamp yes, but there might be the case where you don't any relevant header to sign. As per previous discussion, we said you are supposed to pass only headers you want to sign and the headers object can be empty. 
However I agree, we are eventually always signing the timestamp at least, so that should always be part of the list.